test/e2e-test-runner.js

```diff
@@ -0,0 +1,120 @@
+// test-runner.js
+const { spawn } = require('node:child_process');
+const path = require('node:path');
+const glob = require('fast-glob');
+const { pLimit } = require('plimit-lit');
+const chalk = require('chalk'); // Import chalk for colorizing output
+const { getPublishedSourceCodeIds } = require('./utils'); // Import the function for getting source IDs
+
+// Define the concurrency limit for test runs using `plimit-lit`
+const limit = pLimit(5);
+
+// Define the pattern to match all .test.js files in the current directory and subdirectories
+const testPattern = path.resolve(__dirname, '**/*.test.js');
+
+// Use `fast-glob` to find all matching test files
+const testFiles = glob.sync(testPattern);
+
+// Check if any test files were found
+if (testFiles.length === 0) {
+  console.error('No test files found. Aborting test run.');
+  process.exit(1);
+}
+
+// Retrieve the source code IDs using the previously defined function
+const luaSourceIds = getPublishedSourceCodeIds();
+
+// Check if any source code IDs were found
+if (luaSourceIds.length === 0) {
+  console.error('No Lua Source IDs found. Aborting test run.');
+  process.exit(1);
+}
+
+// Store results for final reporting
+const testResults = {
+  successes: [],
+  failures: [],
+};
+
+// Function to spawn a test process for a given `luaSourceId`
+const runTestForSourceId = (luaSourceId) => {
```

Comment:
interesting way to spin up these tests - why not do a for/each loop in a normal test structure?

Reply:
Well, what I was trying to do here was leverage our existing integration tests, where for each test we migrate from a previous source and then perform the integration tests on the migrated source.

We COULD achieve this in a suite with a loop, however that would require portability of the integration tests, or a lot of copy pasta. This seemed the DRY-ist way to achieve that.

As a bonus we can have some speed with conccurently running children if desired, and fresh sandboxed env's on each one. Frankly I don't like this pattern very much since it actually performs a network request for the source on each integration test - I attempted to do some async pre-fetching in the runner and struggled to get that working well.

We could change this to a single e2e file that calls the target integration file with argv for the lua id maybe, then use a suite instead of `run` for the e2e test.